packages/react-app/src/hooks/ContractLoader.js

```diff
@@ -18,59 +18,110 @@ import { useState, useEffect } from "react";
+  - externalContracts: object with chainIds as keys, with an array of contracts for each
 */
 
-const loadContract = (contractName, signer) => {
-  const newContract = new Contract(
-    require(`../contracts/${contractName}.address.js`),
-    require(`../contracts/${contractName}.abi.js`),
-    signer,
-  );
-  try {
-    newContract.bytecode = require(`../contracts/${contractName}.bytecode.js`);
-  } catch (e) {
-    console.log(e);
-  }
-  return newContract;
-};
-
-export default function useContractLoader(providerOrSigner) {
+export default function useContractLoader(providerOrSigner, config = {}) {
   const [contracts, setContracts] = useState();
+
+  const customAddressKeys = config.customAddresses && Object.keys(config.customAddresses).join();
+  const customAddressValues = config.customAddresses && Object.values(config.customAddresses).join();
+
   useEffect(() => {
+    let active = true;
+
     async function loadContracts() {
-      if (typeof providerOrSigner !== "undefined") {
+      if (providerOrSigner && typeof providerOrSigner !== "undefined") {
+        console.log(`loading contracts`);
         try {
           // we need to check to see if this providerOrSigner has a signer or not
           let signer;
+          let provider;
           let accounts;
+
           if (providerOrSigner && typeof providerOrSigner.listAccounts === "function") {
             accounts = await providerOrSigner.listAccounts();
           }
 
-          if (accounts && accounts.length > 0) {
+          if (ethers.Signer.isSigner(providerOrSigner)) {
+            signer = providerOrSigner;
+            provider = signer.provider;
+          } else if (accounts && accounts.length > 0) {
             signer = providerOrSigner.getSigner();
+            provider = providerOrSigner;
           } else {
             signer = providerOrSigner;
+            provider = providerOrSigner;
+          }
+
+          const providerNetwork = await provider.getNetwork();
+
+          const _chainId = config.chainId || providerNetwork.chainId;
+
+          let contractList = {};
+          let externalContractList = {};
+          try {
+            contractList = config.hardhatContracts || require("../contracts/hardhat_contracts.json");
+          } catch (e) {
+            console.log(e);
+          }
+          try {
+            externalContractList = config.externalContracts || require("../contracts/external_contracts.js");
+          } catch (e) {
+            console.log(e);
           }
 
-          const contractList = require("../contracts/contracts.js");
+          let combinedContracts = {};
+
+          if (contractList[_chainId]) {
+            for (const hardhatNetwork in contractList[_chainId]) {
+              if (Object.prototype.hasOwnProperty.call(contractList[_chainId], hardhatNetwork)) {
```

Comment:
It looks like this does the same as
```js
for (const hardhatNetwork in Object.keys(contractList[_chainId])) {
  // now we don't need the if because we only iterate on the object keys, not the whole prototype
}
```
I'm not sure if we're doing it this way for some other reason that I might be missing.